parse.y

```diff
@@ -8140,16 +8263,16 @@ yycompile0(VALUE arg)
         tree = NEW_NIL(&NULL_LOC);
     }
     else {
-        VALUE tokens = p->tokens;
+        rb_parser_ary_t *tokens = p->tokens;
         NODE *prelude;
         NODE *body = parser_append_options(p, RNODE_SCOPE(tree)->nd_body);
         prelude = block_append(p, p->eval_tree_begin, body);
         RNODE_SCOPE(tree)->nd_body = prelude;
         p->ast->body.frozen_string_literal = p->frozen_string_literal;
         p->ast->body.coverage_enabled = cov;
         if (p->keep_tokens) {
-            rb_obj_freeze(tokens);
-            rb_ast_set_tokens(p->ast, tokens);
+            p->ast->node_buffer->tokens = tokens;
+            p->tokens = NULL;
```

Comment:
Could you add free for `rb_ruby_parser_free` just in case, if an exception happens in the middle of parsing, this line can be skipped.

Reply:
Is this OK? https://github.com/hasumikin/ruby/commit/c2d3fc8ec4b536f122ddc5a3748145481b2d3385 (I amend a function name `parser_ary_free()` to `parser_token_free()` for clarity)

Reply:
`parser_token_free` is preferred, BTW the existence of `ast` and `ast->node_buffer` is confirmed by `rb_ast_free`. Therefore it's more simple to call `parser_tokens_free` directly in `rb_node_buffer_free`.

Reply:
You are right. Fixed https://github.com/hasumikin/ruby/commit/e9e3ac6436adbbdd50ed39315fc1406508f8f3bd